### ISSUE TITLE
Deployment architecture

### ISSUE THREAD

**Issue by mlissner** [freelawproject/bigcases2#4]
Since we're in an k8s world these days, I thought I'd jot some thoughts about how to move this to prod. Ansel, I'm not sure how you feel about this, but I imagine it feels kind of gross because you're in a happy "Iterate on my laptop" mode, but if we can bring this all into k8s with the rest of our architecture, that's probably best in terms of doing things the right way. It's not terribly difficult, I don't think, but let me know what you think. 

I forget your familiarity with k8s, so let me know if I'm not making sense (but I think you actually know k8s? I forget).

Here are the pieces we probably want to think about:

 - We should move to Postgresql instead of SQLite, at least in prod (SQLite is fine in dev, sure). I think this will just mean having different settings, and Flask should make this a simple abstraction. We have a postgresql DB we can easily use for this. This is needed because in the k8s world, all containers can disappear at any time, so there's no particularly good place to store a SQLite DB. 

    As I write that though, I realize that it's not entirely true. We _could_ put the SQLite DB onto a elastic block storage device. That'd work too, but feels pretty janky. Like, I sort of would expect somebody to delete that EBS device without realizing they were nuking a DB.

    The other thing that doing this implies is that we'll need different settings in prod vs dev, and some way to pick up those settings via (probably) envs. That's a pain, but we've done it for CL (in Django).

 - For ingress of webhook events and any kind of admin panel, we'll need a domain name and some kind of routing. We use Elastic Load Balancers for this in CL, which means we don't need Nginx or Apache. They're cheap, scalable, and relatively easy. They take care of HTTPS too.

    I guess the domain name could live under free.law as a subdomain? bigcases.free.law?

 - For the actual Flask application, we'll want to dockerize it. We can just follow the example of CL, I imagine, and that should work.

 - If we do all this, we'll have the option of putting bcb2 into a different k8s namespace from CL. The con doing that is that it's a bunch of work. The pro is that we would have network and feature isolation that could be useful somehow, someday. I think I lean towards putting it into CL's namespace, just because it's easy.
 
 - If we share CL's namespace, we will have access to doctor and similar microservices. That's nice! 

 - Finally, when all of this is done, we'll probably want to think about automatic deployment, like we have on CL, so that when code is merged, it gets automatically deployed to k8s via Github Actions. That's a really nice feature that makes life so much better and we have a template for it that we should be able to follow. 

There. A k8s migration game plan in half a dozen slightly vague steps! This is a lot of detail, I realize, and I like writing things down for my own purposes, but hopefully this will get thoughts flowing.

**Comment by mlissner** [comment]
(If you're reading my last message via email, I updated it on Github with a few clarifications). 

One other thought on this:

>  The pro [of k8s namespace isolation] is that we would have network and feature isolation that could be useful somehow, someday

I can already think of a security reason that namespace/network isolation would be smart, though I don't think it's enough of a reason to bother with it yet. 

In https://github.com/freelawproject/courtlistener/issues/2273, we're looking at the security risks created by the webhooks system, and one thing that comes to mind now is that if we have BCB2 in the same k8s namespace, somebody could create a webhook in CL that posted to BCB from within the cluster.

All you'd have to do is set up a webhook with the url wherever BCB2 lives in the cluster, and then CL would start sending it POSTs that it wasn't expecting. Because BCB is expecting posts from CL, the POSTs would even be formatted properly and could even be parsed properly by BCB.

(Probably not ideal. Of course, if BCB2 doesn't have security on its webhook endpoints, you could also set up a webhook in CL to send outside the cluster to `bigcases.free.law/webhooks/docket` and cause trouble that way even if the namespaces were isolated, but it seems a bit better at least to force the attack to come from outside the cluster.)

**Comment by anseljh** [comment]
That all sounds pretty reasonable. It detracts a bit from the "API example app" goal--but we can achieve that in some other project. I am not up on k8s at all, or much on Docker, so I'd need help with all this.

**Comment by anseljh** [comment]
@mlissner I switched to Postgres and threw a `Dockerfile` in to spin up a dev Postgres server, which is in https://github.com/freelawproject/bigcases2/tree/main/database.

**Comment by mlissner** [comment]
If it's helpful, the order I'd prioritize this work in would be something like:

1. Get bcb2 into two docker containers. One served by gunicorn in prod and one served by flask's dev server in dev (CL has a good example of this with optimized gunicorn configs).

1. Create a docker compose file for developers that launches postgres, the new dev bcb2 container, and doctor (CL has this too).

2. Make bcb2 use env files for settings and make that work with the docker compose file (for example, in Django, [we use this](https://github.com/joke2k/django-environ), I assume Flask has something similar).

I think with that stuff done, FLP folks can take care of the rest of the infrastructure stuff, I'd say.

**Comment by anseljh** [comment]
Nice. Sounds good. Who can show me the ropes on Doctor?

**Comment by mlissner** [comment]
That would be @flooie, whom I suggest you grab on slack when you have questions. IIRC, Doctor had the ability to make multiple thumbnails at a time for a PDF, but then lost it for some reason, so you might see something saying it works when it doesn't. I did just work on its docs though, so I think they should be in good shape.

**Comment by mlissner** [comment]
I think this is our next priority for BCB. I want it in prod as soon as possible. Then we can iterate.

**Comment by mlissner** [comment]
Thinking about this some more:

> one thing that comes to mind now is that if we have BCB2 in the same k8s namespace, somebody could create a webhook in CL that posted to BCB from within the cluster.

It...won't matter. We'd get webhook events, we'd try to match them to a subscription. That'd fail, and nothing would happen. Folks could cause trouble this way, but it's *really* limited in how much damage it could do.

**Comment by mlissner** [comment]
I'm making some steady progress on this:

 - Our env file is complete and deployed to k8s.
 - Our DB is created in RDS and configured in the deployed env  (2 vCPUs, 500MB RAM, Single Availability Zone)
     - The DB is accessible in the VPC via the security group and the correct IP address allowlisting
 - A horizontal auto-scaler is in place for the python app so that it can scale to eight replicas if needed (this would be weird). By default it has four replicas, which are spread across availability zones in AWS.
 - The pods are deployed in k8s and connecting to their variables and RDS.
 - The database is migrated
 - Gunicorn is working (it was needed as a dependency)

And finally:
 - The healthcheck from #30 is working, I think.

The missing piece is actually putting a webserver (called an "ingress" in k8s parlance) in front of this. We'll see if I can get that going today or tomorrow.

Next step will be to get the ALB set up via kubernetes.

**Comment by mlissner** [comment]
OK, I'm happy to share that https://bots.law is now live and in pretty fantastic shape:

 - It's scalable, secure, reliable, and performant.
 - It automatically deploys when code is merged.
 - It gets an A+ score for it's SSL: https://www.ssllabs.com/ssltest/analyze.html?d=bots.law&latest
 - It gets an A score for it's security headers: https://securityheaders.com/?q=bots.law&followRedirects=on

Go ahead and give it a try! 

The final infrastructure uses:
 - Route 53 as DNS
 - Cloudfront for SSL termination and global point of access (CDN without caching). 
 - AWS Application Load Balancer to serve as a distributed webserver.
 - Amazon Certificate Manager to provide SSL certs to the ALB and Cloudfront.
 - ALB connects to k8s's DNS system that I don't fully understand to get our traffic to a set of dockerized django replicas.

Two things that got me hung up were:
 - When you create the ingress, the Listeners automatically get set up with the www.courtlistener.com certificate. That doesn't work and has to be changed.
 - Sentry issue BIGCASES2-3 was due to setting the SSL mode to `require` on our database connection. It appears to be related to this issue with psychopg2 or maybe with libpq, not sure: https://github.com/psycopg/psycopg2/issues/1535

Anyhow, this is all set, folks. Closing!

**Comment by mlissner** [comment]
It also has access to doctor, selenium, postgresql, and redis. That's a full system, there, folks!

**Comment by mlissner** [comment]
Well, it *was* working! I guess I spoke too soon. Cloudfront is reporting an error, so I'm testing from the pod on up to see what's wrong.

The pod test works fine:

```
curl -k localhost:8888
```

The alb test works fine:

```
curl -v -k -H 'X-ALB-SECRET: XYZ' -H "host: bots.law" https://ALB_URL.us-west-2.elb.amazonaws.com
```

So it had to be cloudfront. One test I began to do was to swap out cloudfront for the ALB in our DNS, but I realized it wouldn't work because the ALB requires a secret header to work. 

I eventually figured out that the listener's certificate somehow got set back to courtlistener.com instead of bots.law. I don't know how that happened. We'll have to see if it happens again.
